### PR TITLE
bpo-34866: Adding max_num_fields to cgi.FieldStorage

### PR DESCRIPTION
**Pull request by matthewbelisle-wf**
Adding `max_num_fields` to `cgi.FieldStorage` to make DOS attacks harder by
limiting the number of `MiniFieldStorage` objects created by `FieldStorage`.

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](https://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for your contribution, we look forward to reviewing it!

File: Lib/cgi.py
Comment (matthewbelisle-wf) [inline Lib/cgi.py:590]: Not an essential change, but the list comprehension seemed faster and cleaner than the `append()`.

File: Lib/cgi.py
Comment (matthewbelisle-wf) [inline Lib/cgi.py:606]: Not an essential change, but the extend with generator seemed faster and cleaner than the `append()`.

File: Lib/urllib/parse.py
Comment (matthewbelisle-wf) [inline Lib/urllib/parse.py]: Doing this with `finditer()` instead of `count()` so the worst case time is bounded by `max_num_fields` instead of `len(qs)`.

**Comment by matthewbelisle-wf** [comment]
@edevil You might be interested in this, I noticed you wrote your own `limited_parse_qsl ()` for django.

https://github.com/django/django/blob/2.1.1/django/utils/http.py#L409-L415

**Comment by edevil** [comment]
@matthewbelisle-wf Ah! This would have helped back then. :)

File: Lib/cgi.py
Comment (tseaver) [inline Lib/cgi.py]: ISTM that we should be propagating through the `max_num_fields` limit for sub-parts (and maybe reducing by the aggregate number of fields read so far).

File: Lib/test/test_cgi.py
Comment (tseaver) [inline Lib/test/test_cgi.py]: Given that there are only 3 parts in `data`, it seems as though this shouldn't raise:  do we have an off-by-one somewhere?

**Review by tseaver — COMMENTED** [review]
I like the overall approach here.

File: Lib/test/test_cgi.py
Comment (matthewbelisle-wf) [reply Lib/test/test_cgi.py]: @tseaver You're right there are three parts in `data`, but there are two extra parts in `QUERY_STRING`. This was to test that both are considered.

File: Lib/cgi.py
Comment (matthewbelisle-wf) [reply Lib/cgi.py]: Good thought, I see what you mean there. Thinking about how to implement that.

Comment (matthewbelisle-wf) [reply Lib/cgi.py]: Okay, added that in 21e5e488a4 along with unit tests.

**Comment by bertjwregeer** [comment]
As the maintainer for WebOb I welcome this change as it will help alleviate one class of attacks.

File: Lib/cgi.py
Comment (matthewbelisle-wf) [inline Lib/cgi.py:647]: @tseaver I'm assuming here that if a user declares a custom `FieldStorageClass` it has to accept the same params as this class. If that isn't a good assumption, aka breaking change, then let me know. So far the other parameters have had the same assumption.

**Comment by alex** [comment]
@methane and @ambv, looks like y'all were some of the most recent folks to touch these files, would you mind taking a look? Thanks!

File: Lib/cgi.py
Comment (ambv) [inline Lib/cgi.py]: If you really want a type annotation (of which I see no precedent above), then just write "int".

File: Lib/cgi.py
Comment (ambv) [reply Lib/cgi.py:590]: This is a nice improvement. Good catch!

File: Lib/urllib/parse.py
Comment (ambv) [inline Lib/urllib/parse.py]: Ditto about the annotation.

**Review by ambv — APPROVED** [review]
LGTM. I'll let @methane have his say and merge.

Thanks! ✨ 🍰 ✨

File: Lib/cgi.py
Comment (matthewbelisle-wf) [reply Lib/cgi.py]: Thanks, fixed in ab0eb93f21.

**Comment by matthewbelisle-wf** [comment]
Not sure why the "Azure Pipelines PR" task is failing, but I don't think it is related to this pull request.

**Comment by tirkarthi** [comment]
It's failing for a lot of PRs. I have added an issue for that : https://bugs.python.org/issue34902

**Comment by ambv** [comment]
The failure indeed looks unrelated:

screenshot 2018-10-07 11 03 53

**Comment by ambv** [comment]
cc @zooba

**Comment by tirkarthi** [comment]
@ambv The issue was resolved with : https://bugs.python.org/issue34902

**Comment by ambv** [comment]
Right. So it looks like @matthewbelisle-wf should just rebase and try again :-)

**Comment by matthewbelisle-wf** [comment]
@methane When you look at this can you add the backport labels so that the [miss-islington bot](https://devguide.python.org/committing/#backporting-changes-to-an-older-version) automatically backports it? I expect the automatic backports to fail on 2.7 but I'll fix that manually.

**Comment by matthewbelisle-wf** [comment]
Also, if I should rebase this to a single commit then let me know @ambv . I did a merge instead of a rebase to keep the comments in sync with the commits.

**Comment by edevil** [comment]
Will this be backported to Python 3.5, 3.6, and 3.7? If so I could probably remove my custom parse_qsl() code from Django.

**Comment by methane** [comment]
@edevil Would you point which code is affected?  It's difficult to estimate how this change will break existing code.

**Comment by ambv** [comment]
@edevil wrote:
> Will this be backported to Python 3.5, 3.6, and 3.7? If so I could probably remove my custom parse_qsl() code from Django.

It would only land in 3.5.7, 3.6.8, and 3.7.2. So unless you're willing to claim only those releases are supported by Django, you will still need your custom handling.

**Review by methane — APPROVED** [review]

**Comment by edevil** [comment]
@methane As far as I can tell this is not a breaking change, since the new parameters have a default value that is the current behaviour.

@ambv Well, the Django docs mention "For each version of Python, only the latest micro release (A.B.C) is officially supported." But that is not for me to decide.

**Comment by matthewbelisle-wf** [comment]
@methane Can you add the backport labels so that the [miss-islington bot](https://devguide.python.org/committing/#backporting-changes-to-an-older-version) automatically backports this? I expect the automatic backports to fail on 2.7 but I'll fix that manually. I intend to backport it to all maintained versions, which I believe are 3.5-3.7 and 2.7.

File: Lib/urllib/parse.py
Comment (methane) [reply Lib/urllib/parse.py]: I feel `qs.count('&') + qs.count(';')` is much simpler and faster.
Is this complexity really needed?

When qs is `'a'*1000 + ';&'`, finditer()'s time is bounded by `len(qs)`

Comment (matthewbelisle-wf) [reply Lib/urllib/parse.py]: @methane I see what you're talking about, for that case you pasted it is indeed bounded by `len(qs)`. Some coworkers and I had a discussion about this earlier. `qs.count('&') + qs.count(';')` is about 5x faster than `finditer()` if it has to do the whole string. [Here](https://gist.github.com/matthewbelisle-wf/c5c1489114061f2e1d4054fd13c2b853) are the numbers we measured. But the issue I had in mind when I chose `finditer()` was [bpo issue 34866](https://bugs.python.org/issue34866) (see the example.py there specifically) which is about 50x faster with `finditer()` in the measurements gist. It pretty much comes down do which scenario you want to protect against, and I think either method is an improvement. If you look over that and still think the `count()` method is better then let me know and I'll change it.

Comment (matthewbelisle-wf) [reply Lib/urllib/parse.py]: @methane Let me know what you think of that explanation and I'll make the switch if you want, whichever way is okay with me.

Comment (methane) [reply Lib/urllib/parse.py]: Receiving 9MB query string is slow already.  I don't think "500x slower than finditer" is not a problem in such case.

There are other methods:

```py
pairs = [s2 for s1 in qs.split('&', max_num_fields) for s2 in s1.split(';', max_num_fields)]
if len(pairs) > max_num_fields:  # len(pairs) may be `(max_num_fields+1)*2`.  But since it's not cgi.FieldStorage, it's not a problem.
    raise ValueError('Max number of fields exceeded')
```

Or you can use the regex to make pairs too:

```py
pairs = _QS_DELIMITER_RE.split(qs, max_num_fields)
if len(pairs) > max_num_fields:
    raise ValueError('Max number of fields exceeded')
```

I think these are simpler and faster than `enumerate` hack.

Comment (matthewbelisle-wf) [reply Lib/urllib/parse.py]: Alright, thanks for taking a look @methane . I changed that to use `count()` which is simplest in commit 1fa59e4efd. If that looks good can you add the backport labels so that the [miss-islington bot](https://devguide.python.org/committing/#backporting-changes-to-an-older-version) automatically backports this? I expect the automatic backports to fail on 2.7 but I'll fix that manually. I intend to backport it to all maintained versions, which I believe are 3.5-3.7 and 2.7.

**Pull request merged**
